axelrod/strategies/axelrod_second.py

```diff
@@ -771,3 +771,41 @@ def strategy(self, opponent: Player) -> Action:
                 return random_choice(0.5)
         else:
             return C
+
+class WmAdams(Player):
+    """
+    Strategy submitted to Axelrod's second tournament by William Adams (K44R),
+    and came in fifth in that tournament.
+
+    Count the number of opponent defections after their first move, call
+    `c_defect`.  Defect if c_defect equals 4, 7, or 9.  If c_defect > 9,
+    then defect immediately after opponent defects with probability =
+    (0.5)^(c_defect-1).  Otherwise cooperate.
+
+    Names:
+
+    - WmAdams: [Axelrod1980b]_
+    """
+
+    name = "WmAdams"
+    classifier = {
+        'memory_depth': float('inf'),
+        'stochastic': True,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def strategy(self, opponent: Player) -> Action:
+        if len(self.history) <= 1:
+            return C
+        number_defects = opponent.defections
+        if opponent.history[0] == D:
+            number_defects -= 1
+
+        if number_defects in [4, 7, 9]: return D
```

Comment:
@gaffney2010 can you confirm which strategy this is? Looking at https://github.com/Axelrod-Python/TourExec/blob/master/src/strategies/k44r.f I don't immediately follow how this implementation corresponds to `k44r`?

Reply:
I can confirm that it's k44r.  I'll say more about the translation...

Reply:
I confirm I get the same fingerprints :) 

I think you've just done a good job refactoring and re writing the code to be much more readable, just got me scratching my head to follow along, so one or two pointers for my own benefit just to confirm would be awesome (even if it's not in the PR itself but just in a comment here).

Reply:
The ones I'm particularly stumped by are the comparisons of `number_defects` to 4, 7 and 9 and also where `number_defects - 9` comes from...

Reply:
Basically every time a defect comes up, it increments this MC variable.  While AM is still a whole number, then MC is either less than AM or equal to AM.  [While less than, line 16 returns C.]  While MC=AM it defects; this happens for the first time at 4 defects [line 11].  As soon as MC exceeds AM, then AM is halved, and MC is reset to zero.  [The 5th defect resets MC and sets AM to 2. So the 7th defect gets MC=AM.  Then the 8th defect resets MC and sets AM to 1.  So the 9th defect gets MC-AM]

After the 9th defect, the function changes because the AM is no longer whole, so MC never again equal AM (and line 17 is out of play).  So as soon as MC is incremented (any defect), lines 19-22 are triggered.  [Line 22 was 100% until now.]  And MC is immediately reset, so there's no sustained defecting, like there was earlier.  [For example, there could be exactly 4 defects for many turns, and the strategy would defect for all those turns.  But after the 100th defect, the strategy will defect (with some probability) ONLY once.]

Reply:
That's awesome and make sense. Thanks for explaining.

Reply:
`number_defects - 9` just makes the AM follow the right halving pattern.  AM starts at 4.0 and halves at 4th, 7th, and 9th defects, for AM=1.0.  After that EACH defect will trigger a halving.